jabgui/src/main/java/org/jabref/gui/externalfiles/AutoLinkFilesAction.java

```diff
@@ -61,18 +63,21 @@ public void execute() {
         final NamedCompoundEdit nc = new NamedCompoundEdit(Localization.lang("Automatically set file links"));
 
         Task<AutoSetFileLinksUtil.LinkFilesResult> linkFilesTask = new Task<>() {
-            final BiConsumer<LinkedFile, BibEntry> onLinkedFile = (linkedFile, entry) -> {
+            final BiConsumer<List<LinkedFile>, BibEntry> onLinkedFilesUpdated = (newLinkedFiles, entry) -> {
                 // lambda for gui actions that are relevant when setting the linked file entry when ui is opened
-                String newVal = FileFieldWriter.getStringRepresentation(linkedFile);
+                String newVal = FileFieldWriter.getStringRepresentation(newLinkedFiles);
                 String oldVal = entry.getField(StandardField.FILE).orElse(null);
                 UndoableFieldChange fieldChange = new UndoableFieldChange(entry, StandardField.FILE, oldVal, newVal);
                 nc.addEdit(fieldChange); // push to undo manager is in succeeded
-                UiTaskExecutor.runInJavaFXThread(() -> entry.addFile(linkedFile));
+
+                // Wait because there are several rounds in one auto-link operation
+                // The later round depends on the updated bibEntry of the previous round
+                UiTaskExecutor.runAndWaitInJavaFXThread(() -> entry.setFiles(newLinkedFiles));
             };
```

Comment:
The original `onLinkedFile` is not reasonable, considering that we not only add new linked files, but also update old ones, and multiple linked files may be updated at the same time. So it is changed to `onLinkedFilesUpdated`.